src/main/kotlin/org/rust/lang/core/lexer/RustLiteralLexer.kt

```diff
@@ -0,0 +1,259 @@
+         */
+        fun forRawStringLiterals(): RustLiteralLexer = RustRawStringLiteralLexer()
+
+        /**
+         * Create an instance of [RustLiteralLexer] suitable for given [IElementType].
+         *
+         * For the set of supported token types see [LITERALS_TOKEN_SET].
+         *
+         * @throws IllegalArgumentException when given token type is unsupported
+         */
+        fun of(tokenType: IElementType): RustLiteralLexer = when (tokenType) {
+            INTEGER_LITERAL,
+            FLOAT_LITERAL           -> forNumericLiterals()
+
+            CHAR_LITERAL,
+            BYTE_LITERAL            -> forCharLiterals()
+
+            STRING_LITERAL,
+            BYTE_STRING_LITERAL     -> forStringLiterals()
+
+            RAW_STRING_LITERAL,
+            RAW_BYTE_STRING_LITERAL -> forRawStringLiterals()
+
+            else                    -> throw IllegalArgumentException("unsupported literal type")
+        }
+    }
+}
+
+private class RustNumericLiteralLexer : RustLiteralLexer() {
+    override fun determineTokenType(): IElementType? =
+        if (tokenStart >= tokenEnd) {
+            null
+        } else when (state) {
+            IN_VALUE  -> LiteralTokenTypes.VALUE
+            IN_SUFFIX -> LiteralTokenTypes.SUFFIX
+            else      -> error("unreachable")
+        }
+
+    override fun locateToken(start: Int): Int =
+        if (start >= bufferEnd) {
+            start
+        } else when (nextState) {
+            IN_VALUE  -> locateValue(start)
+            IN_SUFFIX -> Math.max(start, bufferEnd)
+            else      -> error("unreachable")
+        }
+
+    private fun locateValue(start: Int): Int {
+        var hasExponent = false
+        val (vstart, digits) = when {
+            bufferSequence.startsWith("0b", start) -> (start + 2) to BIN_DIGIT
+            bufferSequence.startsWith("0o", start) -> (start + 2) to OCT_DIGIT
+            bufferSequence.startsWith("0x", start) -> (start + 2) to HEX_DIGIT
+            else                                   -> start to DEC_DIGIT
+        }
+
+        for (i in vstart..(bufferEnd - 1)) {
+            val ch = bufferSequence[i]
+
+            if (!hasExponent && EXP_CHARS.contains(ch)) {
+                hasExponent = true
+                continue
+            }
+
+            if (!digits.contains(ch) && !NUM_OTHER_CHARS.contains(ch)) {
+                nextState = IN_SUFFIX
+                return i
+            }
+        }
+
+        return bufferEnd
+    }
+
+    companion object {
+        // States
+        const val IN_VALUE = 0
+        const val IN_SUFFIX = 1
+
+        // Utility charsets
+        const val DEC_DIGIT = "0123456789"
+        const val BIN_DIGIT = "01"
+        const val OCT_DIGIT = "01234567"
+        const val HEX_DIGIT = "0123456789abcdefABCDEF"
+        const val NUM_OTHER_CHARS = "+-_."
+        const val EXP_CHARS = "eE"
+    }
+}
+
```

Comment:
Here begins a part I'm not so proud of. Hope there aren't any hidden bugs. I'll try to clean this up a bit, but cannot promise anything.